### ISSUE TITLE
Can I create a env based on multiple yaml file?

### ISSUE THREAD

**Issue by chantisnake** [conda/conda#5821]
I want to test my app with multiple python version, therefore we need to create a some env files.

But our requirements are the same on different python version.

Therefore we currently have the following file:
```
- dep_dev  # the dev dependency, like pytest
- dep_user  # the dependency for user, like scikit-learn
- env_main  # the main environment, contains the name and the python version 3.6
- env_py35  # the main environment, contains the name and the python version 3.6
```

what I want to do is like:
```bash
conda create --file .\conda_envs\env_main.yml --file .\conda_envs\dep_dev.yml --file .\conda_envs\dep_user.yml
```

Is there a way I can do that?

**Comment by DavidMertz** [comment]
ISTM that this request is not conceptually possible to fulfill. If, e.g. I have `env1.yml` and `env2.yml`

```yaml
name: py3
dependencies:
  - python=3
```

And:

```yaml
name: py2
dependencies:
  - python=2
```

These cannot be merged automatically.  Even if they did not specify incompatible tool versions, the *name* of the environment is different in each file.

I believe this Issue should be adjusted instead to request the creation of a separate `merge-conda-env` tool that guided creation of a compatible superset environment.  Of course, such a tool does not necessarily need to live in this repo, it could be independent.

**Comment by chantisnake** [comment]
No, each environment has only one file specify name, and there is no conflict in version

**Comment by chantisnake** [comment]
here is approximately what in `env_main`:
```yaml
name: main_env
dependencies:
  - python=3.6
```
here is approximately what in `user_dep`:
```yaml
dependencies:
  - scipy
  - panda
```
here is approximately what in `dev_dep`:
```yaml
dependencies:
  - pytest
  - flake8
```

what I want is:
```
conda create --file .\conda_envs\env_main.yml --file .\conda_envs\dep_dev.yml --file .\conda_envs\dep_user.yml
```

**Comment by DavidMertz** [comment]
My point is that while there are *some* environment files that won't conflict with each other, that's not the rule. What you need is a tool to merge yaml files with suitable checking and decision flags. In your case, this would succeed with no special switches and no errors, but not universally.

**Comment by chantisnake** [comment]
So there is no way that conda will do that natively? I have to merge the file by myself?

**Comment by chantisnake** [comment]
So there is no way that conda will do that natively? I have to merge the file by myself

**Comment by kalefranz** [comment]
Yes. And, as David explained, this is a feature that would be technically very difficult to implement correctly. Sorry for the bad news.

**Comment by uniqueg** [comment]
I don't at all understand why this should be so difficult to implement. Sure, there may be conflicting dependencies if the user provides more than one YAML file. As there may be if only one file is provided. What's the difference?

If the merging of the provided files happens before anything is resolved/installed, package conflicts could be identified and processed according to some rules (e.g., exit with warning or, better/optionally, have files override previously specified packages in the order they appear on the command line, which is pretty standard). The same would work for the `name` field as well. Overriding conflicting values in this manner is pretty standard behavior for configs. Compare, e.g., [`docker-compose`](https://docs.docker.com/compose/reference/overview/#specifying-multiple-compose-files) among many others.

**Comment by mrclary** [comment]
I agree with @uniqueg . `conda create` allows multiple `--file` specifications, as long as `name: <name>` is not present in any of them. `conda env create` only allows one `--file` specification, which may or may not have `name: <name>` present.

One complaint I have had for a long time is that conda has two methods for creating an environment and they each behave differently; I always forget which one behaves which way and its very frustrating. I think that there should only be one method for creating an environment, and it _should_ accept multiple `--file` specifications.
Files should combine as follows:
* package names unioned, e.g. `pkg1 pkg2` + `pkg2 pkg3` -> `pkg1 pkg2 pkg3`
* package versions intersected, e.g. `pkg > 1.0` + `pkg < 2.0` -> `pkg >1.0,<2.0`
* incompatible intersections raise error or first encounter given precedence
* The first `name: <name>` encountered should be given precedence
* The option `-n <name>` should be given precedence over any file spec

**Comment by AlexanderCecile** [comment]
> I agree with @uniqueg . `conda create` allows multiple `--file` specifications, as long as `name: <name>` is not present in any of them. `conda env create` only allows one `--file` specification, which may or may not have `name: <name>` present.


Doesn't the `file` argument in `conda create` expects a different format than the `file` argument in `conda env create`, though?

**Comment by AlexanderCecile** [comment]
> I agree with @uniqueg . `conda create` allows multiple `--file` specifications, as long as `name: <name>` is not present in any of them. `conda env create` only allows one `--file` specification, which may or may not have `name: <name>` present.


Doesn't the `file` argument in `conda create` expect a different format than the one in `conda env create`, though?

**Comment by mrclary** [comment]
> Doesn't the `file` argument in `conda create` expect a different format than the one in `conda env create`, though?

Hence my comment:
> One complaint I have had for a long time is that conda has two methods for creating an environment and they each behave differently; I always forget which one behaves which way and its very frustrating.

I would also add that ``conda create -h`` and ``conda env create -h`` do not specify the supported file formats.

**Comment by mrclary** [comment]
Okay, perhaps I was a bit too hasty. The help does use the terminology "Create a new conda environment from a list of specified packages" for `conda create` and "Create an environment based on an environment file" for `conda env create`. If I interpret this to mean that `conda create` _cannot_ take yml file format and that `conda env create` _only_ takes a yml format, then the different behaviors at least have a little more rhyme, even if still lacking in reason.

Following is my summary of basic usage...

**Test Files**
<details><summary>test.yml</summary>

```
name: test
channels:
  - defaults
dependencies:
  - numpy=1.18.1
  - python=3.7
```
</details>
<details><summary>test-a.yml</summary>

```
name: test
channels:
  - defaults
dependencies:
  - python=3.7
```
</details>
<details><summary>test-b.yml</summary>

```
name: test
channels:
  - defaults
dependencies:
  - numpy=1.18.1
```
</details>
<details><summary>test.txt</summary>

```
numpy=1.18.1
python=3.7
```
</details>
<details><summary>test-a.txt</summary>

```
python=3.7
```
</details>
<details><summary>test.txt</summary>

```
numpy=1.18.1
```
</details>

**Example 1**
```bash
$ conda env create --file test.yml
```
Works just fine.

**Example 2**
```bash
$ conda env create --file test-a.yml --file test-b.yml
```
Ignores test-a.yml, but otherwise does not fail.

**Example 3**
```bash
$ conda create -n test --file test.txt
```
Works just fine.

**Example 4**
```bash
$ conda create -n test --file test-a.txt --file test-b.txt
```
Works just fine.

In short, I believe that the behavior sought by the OP can be achieved, but _only_ by using specification files and _not_ yaml files. The practicality depends on further implementation details: e.g. do you plan to re-build the same environment? do you require different channels? etc.
* `conda env create` allows `--force` optional argument to re-build an existing environment; `conda create` with a spec file(s) requires you to remove an existing environment.
* conda channels can be specified in the yaml file with `conda env create`, but have to be provided as command-line options with `conda create`

**Comment by AlexanderCecile** [comment]
> > Doesn't the `file` argument in `conda create` expect a different format than the one in `conda env create`, though?
> 
> Hence my comment:
> 
> > One complaint I have had for a long time is that conda has two methods for creating an environment and they each behave differently; I always forget which one behaves which way and its very frustrating.


Hmm, looking back I believe I had misunderstood part of your comment!
